날씨 앱 [STEP 3-2] Blu, Jenna

calledBlu opened the pull request:
>안녕하세요! 코리(@corykim0829)🐶!
🌙**TEAM LuNna**🌙 의 Blu, Jenna입니다!
>
>5주라는 시간이 훌쩍 지나갔네요!
>저번 STEP에서 진행했던 UI 구현에 이어 나머지 요구사항을 구현해 보았어요.
>STEP 3-2를 위해 남겨주신 리뷰도 적극! 반영해 보았답니다!
>배경 이미지를 적용하기 위해 시간을 많이 사용했지만 리팩토링을 진행하는 작업이 대부분이라 전 STEP보다는 빠르게 마무리할 수 있었습니다.
>
>정성스럽고 세심한 리뷰 덕분에 많이 성장했습니다! 프로젝트 기간동안 정말 감사했습니다! 🙇🏻🙇🏻

<br>

# 🧑🏼‍💻 작업 목록

## UI 구현(2)
- [x] UICollectionViewLayout의 활용
- [x] Configuration의 활용
    - cell의 `backgroundConfiguration`, `contentConfiguration` 프로퍼티 활용
- [x] 배경 이미지 적용
- [x] Refresh Control을 활용하여 화면을 위에서 아래로 끌었다 놓는 경우 정보 새로고침
- [x] WeatherParser의 Generic 타입에 관한 리팩토링 진행
- [x] 다양한 오류 상황에 대한 에러 타입 정의
- [x] URLComponent 관련 코드 리팩토링
- [x] CollectionView의 가시성 향상을 위하여 파일 분리



<br>

<img src="https://user-images.githubusercontent.com/71758542/231084419-bb872269-c102-4609-adda-1d0907f70224.gif">
<br>


# 📚 학습 키워드 및 학습 내용
## 1. UICollectionViewCompositionalLayout
> 🖇️　참고한 공식문서: 
> - [UICollectionViewCompositionalLayout](https://developer.apple.com/documentation/uikit/uicollectionviewcompositionallayout)
> - [UICollectionLayoutListConfiguration](https://developer.apple.com/documentation/uikit/uicollectionlayoutlistconfiguration)
> - [UICollectionViewListCell](https://developer.apple.com/documentation/uikit/uicollectionviewlistcell)
> - [UIBackgroundConfiguration](https://developer.apple.com/documentation/uikit/uibackgroundconfiguration)
- **학습 내용**
   - `UICollectionViewCompositionalLayout`
       - Collection View 레이아웃의 한 종류
       - 적응성이 높고 유연한 배열로 아이템을 결합할 수 있는 레이아웃 개체
       - `list(using:)` 메서드 사용 시 지정된 구성의 list 섹션만 포함하는 레이아웃을 구성
       - 위 메서드를 사용하기 위해서는 `UICollectionLayoutListConfiguration`으로 layout을 구성해 주어야 함
    - `UICollectionLayoutListConfiguration`
        - list의 header와 footer를 설정하기 위해 필요한 구조체
        - `headerMode`, `footerMode` 프로퍼티를 .supplementary로 설정하여 보조 view를 사용하여 header, footer 표시 가능
        - 위 프로퍼티들은 .none이 기본값, `headerMode`의 경우 .firstItemInSection으로 Section 내의 첫 번째 아이템을 헤더로 사용할 수 있음
        - `backgroundColor`는 배경색을 완전히 없애기 위한 프로퍼티
            - 기본 값은 nil이지만, 이는 system background color를 사용한 것 → 우리 프로젝트에서는 white로 지정됨
        - appearance로 list의 모양을 지정해줄 수 있음
    - `UICollectionViewListCell`
        - list 기능과 기본 스타일을 사용하기 위해 사용되는 Collection View cell 클래스
        - list의 셀 액세서리나 상호작용 동작 등을 지원하는 타입
    - `UIBackgroundConfiguration`
        - View에 대해 배경을 구성하는 타입
        - UIButton에 직접 적용하거나 UICollectionView/UITableView의 cell, header, footer에 적용할 수 있음
- **적용 과정**
   - FlowLayout에서 CompositionalLayout으로 리팩토링하는 과정에서 `UICollectionViewCell` 대신 `UICollectionViewListCell`으로 대체하려고 하였으나 현재 프로젝트 내에서는 list의 기능을 사용할 일이 없으므로 학습만 진행
- **문제와 해결**
   - View의 backgroundColor를 .clear로 설정하였으나 배경색이 사라지지 않는 문제 발생
   - View를 구성하는 configuration의 backgroundColor를 .clear로 해주어야 배경색이 사라짐
- **아쉬운 점**
   - `defaultContentConfiguration()`을 활용하여 셀을 구성해보지 못한 것이 조금 아쉬움

<br>

## 2. URLComponents
> 🖇️　참고한 공식문서: 
> - [URLComponents](https://developer.apple.com/documentation/foundation/urlcomponents)
- **적용 과정**
   - **기존)**
   날씨 데이터/아이콘 중 무엇을 불러오기 위한 url이냐에 따라 URL 구성요소가 달라져 매개변수가 다른 메서드를 여럿 구현하는 과정에서 중복표현이 다수 발생
   - **수정)**
      - 기존에는 path, queryItems 프로퍼티만 활용했지만, 이번에는 scheme, host 등 다른 url요소도 적극적으로 다뤄보기로 함
      - 데이터/아이콘을 불러오기 위한 메서드 각각에서 path, queryItems를 직접 할당해줬던 기존 코드가 OCP에 좋지 않다고 판단, 불러올 타입에 따라 여러 가지 url요소들(scheme, host, path, query 등)을 계산속성으로 알려주는 WeatherRouter 열거형을 구현
      - URL 및 URLRequest를 만드는 기능은 연관성이 깊은 WeatherRouter에 맡기고, 실제로 데이터를 받아오는 기능만 WeatherParser에 구현
- **문제와 해결**
   |트러블슈팅 A|제네릭타입 명시 문제|
   |:-:|:-|
   |😰|- 불러올 타입에 따른 중복로직을 최대한 통합하려고 했으나, 어디까지나 url생성 및 변환을 위한 정보를 매개변수를 통해 전달해줘야 하므로 제네릭타입 관련 문제점이 발생<br>- WeatherParser에서 날씨데이터가 아닌 날씨아이콘을 parse하는 메서드는 WeatherParser의 제네릭타입 T의 정보를 필요로 하지 않음에도 호출부에서 제네릭 타입을 넣어줘야 했으므로 불필요&적절치 않은 비용을 발생시킴|
   |🚀|- WeatherParser타입이 아닌, 날씨데이터를 불러오는 메서드에만 단독으로 제네릭타입을 붙이기로 함|
   
   |트러블슈팅 B|제네릭타입 추론 문제|
   |:-:|:-|
   |😰|- 제네릭 메서드의 호출부에서 제네릭타입을 추론하기 위한 단서를 요구<br>- 메서드의 반환형을 담을 상수에 타입을 명시하려 했으나, 협업인원이 늘어나고 이러한 유의사항의 공유가 활발하지 않다면 이를 놓쳐 휴먼에러가 발생할 여지가 크다는 우려 발생
   |🚀|- 제네릭타입 T의 추론을 위한 단서(`T.Type`)를 해당 메서드의 매개변수로 받도록 강제함으로써 위의 우려를 해결|
- **아쉬운 점**
   - host 프로퍼티의 경우 맨앞의 "api." 외에는 모두 동일함
   - 아래 3가지를 고민하다가 3번을 택했지만, 여전히 중복표현이 많아 좋은 방법인지 확신이 서지 않음
       1. 'api.'를 제외한 부분(openweathermap.org)을 공통상수로 선언 후, 한쪽에만 'api.'를 붙여 host에 할당하는 방법, 
       2. `subhost: String` 계산속성을 만들어 한쪽은 "api.", 한쪽은 ""를 넣은 후 host = subhost + host 해주는 방법, 
       3. 한쪽은 "api.open~org", 한쪽은 "open~org"을 host에 할당하는 방법
     

<br>

## 3. UIRefreshControl
> 🖇️　참고한 공식문서: 
> - [UIRefreshControl](https://developer.apple.com/documentation/uikit/uirefreshcontrol)
- **학습 내용**
   - `UIRefreshControl` 인스턴스에 `addTarget(_:action:for:)`로 action method를 더해주면 target은 refreshControl의 beginRefreshing을 감지하고 controlEvents를 위해 selector안에 넣어 둔 objc메서드를 실행함
   - `UICollectionView` 인스턴스는 `UIRefreshControl?`타입의 `refreshControl`프로퍼티를 가짐
   (따라서 기본값이 nil이므로 새로운 UIRefreshControl 인스턴스를 만들어 addTarget한 후 할당해줘야 함)
- **적용 과정**
   - `configureRefreshControl()` 메서드 및 objc메서드 `refreshCollectionView()`를 ViewController에 구현

<br>

ueunli committed:
[refactor] 콜렉션뷰 UI요소타입들의 실패가능생성자 내부로직 보강

- `init(frame:)`에서 호출하는 메서드를 `init?(coder:)`에서도 동일하게 호출하도록 수정

calledBlu committed:
[refactor] String extention에 존재하던 주소 변환 초기화자를 메서드로 분리

- CLPlacemark extension의 formatAddress() 메서드로 분리

ueunli committed:
[refactor] API network 및 url 관련 타입 및 로직 수정

- url에 필요한 매직스트링을 String extension으로 분리
- 네이밍 변경) `URLService` → `WeatherRouter`
- `WeatherRouter`에 요청의 종류에 따라 URL 요소들을 저장·조합하는 계산속성 정의
- `WeatherParser` 타입 자체가 아닌 특정 메서드에 제네릭을 적용, 해당 메서드의 매개변수를 추가하여 제네릭타입을 추론하도록 리팩토링

calledBlu committed:
[refactor] Collection View 레이아웃 변경 및 셀 타입 변경, 배경 이미지 추가

- UICollectionViewFlowLayout → UICollectionViewCompositional.list(using: )
- UICollectionViewFlowLayoutDelegate, UICollectionViewDelegateFlowLayout 삭제
- CollectionView의 backgroundView를 UIImageView로 설정
- ForecastWeatherCell과 CollectionView의 configuration을 활용하여 .clear로 지정
- 배경 이미지에 해당하는 파일을 Asset에 추가

ueunli committed:
[refactor] 정보를 새로고침하는 기능 구현 및 ViewController 파일 분리

- collectionview의 refreshControl 할당
- ViewController가독성 개선을 위한 파일 분리

calledBlu committed:
[design] 접근성을 높이기 위해 레이블 스타일 변경 및 cell 내부 요소의 제약조건 수정

ueunli committed:
[design] 날씨아이콘 가시성 향상을 위해 테두리 추가

- UIImageView요소의 layer프로퍼티를 수정하여 얇은 shadow 추가

calledBlu committed:
[refactor] collectionView cell의 타입 변경

- list 기능을 사용하지 않으므로 UICollectionViewListCell → UICollectionViewCell 타입 변경

ueunli committed:
[chore] 불필요한 데이터 삭제

- 더미데이터 `CurrentWeather.json`, `FiveDaysWeather.json` 삭제

corykim0829 commented on WeatherForecast/WeatherForecast/Controller/ViewController+Location.swift:
```diff
@@ -0,0 +1,79 @@
+    }
+    
+    func locationManager(_ manager: CLLocationManager, didFailWithError error: Error) {
+        print(WeatherNetworkError.requestFailed("위치").description)
```
프로덕트 코드에는 print를 넣지 않고, 실패 시 예외 처리를 해주어야합니다! 참고하세요 😊

corykim0829 commented on WeatherForecast/WeatherForecast/Controller/ViewController.swift:
```diff
@@ -9,148 +9,61 @@ import CoreLocation
-        try await currentWeather?.convertToImage {
-            self.currentWeather?.iconImage = $0
-        }
+    private func configureRefreshControl() {
```
collectionview의 refreshControl 사용 좋네요! 👍

corykim0829 commented on WeatherForecast/WeatherForecast/Controller/ViewController.swift:
```diff
@@ -9,148 +9,61 @@ import CoreLocation
-        }
+    @objc private func refreshCollectionView() {
+        locationManager.requestLocation()
+        collectionView.refreshControl?.endRefreshing()
```
여기서 바로 endRefreshing 함수를 호출하면, 실제 요청의 응답은 오지않았는데, refresh가 끝난 느낌이 들어서 어색한 경험을 주는 것 같아요.
완료 시점에 endRefreshing을 처리해보는 것도 고려해보세요!

ueunli replied:
`collectionView`의 `reloadData()`까지 된 후에 `endRefreshing()`하도록 수정했어요!
refreshing의 의미나 수반되는 비동기 작업에 대해 충분히 고려하지 못해 놓친 부분이었는데 짚어 주셔서 감사합니다💚

> 해당 커밋: ec11577

corykim0829 commented on WeatherForecast/WeatherForecast/Model/Extension/String+.swift:
```diff
@@ -6,9 +6,11 @@
-import CoreLocation
 
 extension String {
+    static var weatherMeasurementUnit: Self { "metric" }
```
특정 클래스에서만 사용될 것 같은 상수네요! String 타입 자체에 정적 변수로 선언하기 보다는 해당 클래스에서 중첩 enum으로 처리하면 더 깔끔할 것 같네요!

calledBlu replied:
중첩 enum 사용에 대해 고민하다가 extension으로 구현하게 되었는데 코리가 알려주신 방법이 더 깔끔하게 느껴져서 수정하여 반영해 보았습니다! 감사합니다 ☺️

> 해당 커밋: 23ccf86

corykim0829 commented on WeatherForecast/WeatherForecast/Controller/ViewController+Location.swift:
```diff
@@ -0,0 +1,79 @@
+        let coordinate = CurrentCoordinate(of: location)
+
+        Task {
+            updateAddress(to: location) {
+                $0.formatAddress()
+            }
```
이 부분이 한번에 봤을 때 이해하기 힘든 코드 같아요.
또한 async도 아니니 굳이 Task안에 들어가있을 필요가 없을 것 같네요

updateAddress이니 굳이 completion을 통해 복잡하게 여기서 userAddress를 업데이트를 하기 보다는 
해당 함수 안에서 userAddress를 업데이트해주는게 더 깔끔할 것 같네요

completion을 사용하고 싶다면, completion에서 값만 넘겨주고 userAddress 실제 값을 넣어주는 부분은 여기 completion block 안에서 처리하는게 더 자연스러운 것 같아요!

calledBlu replied:
더 깔끔하게 처리할 수 있는 방법 알려주셔서 감사해요! 해당 메서드 안에서 `formatAddress()`를 수행하도록 변경하였습니다!

> 해당 커밋: 90b1d70

corykim0829 commented on WeatherForecast/WeatherForecast/Controller/ViewController+Location.swift:
```diff
@@ -0,0 +1,79 @@
+            }
+
+            try await updateCurrentWeather(for: coordinate)
+            try await updateForecastWeather(for: coordinate)
```
에러를 예외 처리하는 부분이 빠져있는 것 같아요!
항상 예외 처리를 고민해야합니다!

corykim0829 commented on WeatherForecast/WeatherForecast/Controller/ViewController+Location.swift:
```diff
@@ -0,0 +1,79 @@
+        print(WeatherNetworkError.requestFailed("위치").description)
+    }
+    
+    private func updateCurrentWeather(for location: CurrentCoordinate) async throws {
```
for보다는 at이 좀 더 자연스러운 네이밍일 것 같아요. 참고만하세요:)

calledBlu replied:
해당 부분 참고하여 수정하였습니다! 자연스러운 별칭 추천 감사드려요!

> 해당 커밋: bbbf41e, fe244b1

corykim0829 commented on WeatherForecast/WeatherForecast/Controller/ViewController+Location.swift:
```diff
@@ -0,0 +1,79 @@
+    }
+    
+    private func updateCurrentWeather(for location: CurrentCoordinate) async throws {
+        let current = try await WeatherParser.parseData(at: location, type: CurrentWeatherComponents.self)
```
저번 피드백에서도 말씀드렸었는데, Parser는 파싱을 하는 역할일 것 같은데 이 객체는 네트워크 요청까지 하고 있기 때문에 조금 어색해보이네요!

ueunli replied:
request 및 parse 기능을 포괄할 수 있도록 `WeatherNetworkService`로 네이밍 변경했습니다! 조언 감사해요😆

> 해당 커밋: fe244b1

corykim0829 reviewed (changes requested):
정말 수고 많으셨습니다 👏👏👏👏

전반적으로 깔끔하게 잘 구현하셨고,
UICollectionLayoutListConfiguration도 잘 사용하셔서 구현 잘 하셨습니다 👍
async await도 잘 사용해서 구현하신 부분이 좋습니다 👍

추가적으로
`ForecastWeatherCell`, `WeatherHeaderView`의 prepareForReuse에서 label들만 nil처리로 초기화 해주는데, 이미지도 nil처리를 해주시면 좋을 것 같아요!

또한 예외 처리도 항상 생각해주시면서 개발해주시면 더더욱 좋을 것 같습니다.

수고 많으셨어요 😊

calledBlu committed:
[refactor] `updateAddress(to: )` 내부 로직 수정

- 주소의 형태를 변경하는 메서드 `formatAddress()`를 completion을 사용하지 않고 내부 로직에서 실행되도록 수정

ueunli committed:
[refactor] collectionView의 refreshControl이 endRefreshing()하는 시점을 수정

- locationManager에 지역 업데이트를 요청한 직후가 아니라 collectionView의 reloadData() 완료 직후에 refreshing이 끝나도록 수정

calledBlu committed:
[chore] 메서드의 매개변수 별칭 변경

- updateCurrentWeather(for: )에서 매개변수 별칭 for를 at으로 변경

ueunli committed:
[chore] 매개변수별칭 및 네이밍 일부 변경

- updateForecastWeather(for:) → updateForecastWeather(at:)
- WeatherParser → WeatherNetworkService

calledBlu committed:
[refactor] API URL 생성에 필요한 값들을 WeatherRouter 내의 중첩 열거형으로 구현

- String extension의 static 프로퍼티들을 사용하는 타입 내로 옮기기 위해 삭제
- WeatherRouter 내부의 QueryItemValue으로 구현

corykim0829 commented on WeatherForecast/WeatherForecast/Controller/ViewController+DataSource.swift:
```diff
@@ -0,0 +1,34 @@
+//  Created by J.E on 2023/04/11.
+//
+
+import Foundation
+import UIKit
```
Foundation과 UIKIt을 둘다 import 할 필요는 없습니다! 알아보시고 하나만 import 하면 좋을 것 같네요!

corykim0829 reviewed (approved):
수고 많으셨습니다 😊

라이브러리 import를 전반적으로 정리하면 좋을 것 같네요!
